### ISSUE TITLE
APIs such as @RegisterForReflection should strongly encourage conditionality

### ISSUE THREAD

**Issue by Sanne** [quarkusio/quarkus#20331]
### Description

Follow up on https://github.com/quarkusio/quarkus/issues/20300#issuecomment-924696400

When registering symbols for reflective access, it would be far better to register them based on some condition of reachability.

For example the code actually performing the reflective call: if that code is not reachable (for whatever other reason), then the reflective call is also actually unnecessary.

We've seen this often leads to significant side-effects, as multiple careless reflective registrations make the dead code analysis far less effective.

We should allow expressing such a conditional on all our APIs (including but not exclusively via `@RegisterForReflection` ), and think of strategies to encourage using such conditionals - at least for extension developers.

**Comment by stuartwdouglas** [comment]
I don't think this is particularly useful. @RegisterForReflection is for users to use in their applications, and if application code is not reachable then something is badly wrong.

**Comment by Sanne** [comment]
Ok @stuartwdouglas , I'll close it then; I still want us to strongly encourage extension developers to use conditionals though - and ideally end users should be able to live without using the `@RegisterForReflection` altogether so yes I agree that this is of low value, and possibly not very practical.
